File: q1458/q1458-bottom-up.cpp
class Solution
{
public:
    int maxDotProduct(vector<int> &nums1, vector<int> &nums2)
    {
        vector<vector<int>> dp(nums1.size(), vector<int>(nums2.size()));
        for (int i = 0; i < nums1.size(); i++)
        {
            for (int j = 0; j < nums2.size(); j++)
            {
                dp[i][j] = nums1[i] * nums2[j];
                if (i > 0 && j > 0)
                    dp[i][j] += max(0, dp[i - 1][j - 1]);
                if (i > 0)
                    dp[i][j] = max(dp[i - 1][j], dp[i][j]);
                if (j > 0)
                    dp[i][j] = max(dp[i][j - 1], dp[i][j]);
            }
        }
        return dp[nums1.size() - 1][nums2.size() - 1];
    }
};

File: q1458/q1458-bottom-up.py
class Solution:
    def maxDotProduct(self, nums1, nums2):
        n, m = len(nums1), len(nums2)
        dp = [[float('-inf')] * (m + 1) for _ in range(n + 1)]

        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                take = nums1[i] * nums2[j]
                
                # opt1 = max(take, take + dp[i + 1][j + 1])
                # opt2 = dp[i+1][j]
                # opt3 = dp[i][i-1]
                # dp[i][j] = max(opt1, opt2, opt3)
                dp[i][j] = max(
                    take,
                    take + dp[i + 1][j + 1],
                    dp[i + 1][j],
                    dp[i][j + 1]
                )

        return dp[0][0]
